mmengine/infer/infer.py

```diff
@@ -0,0 +1,468 @@
+        return collate_fn  # type: ignore
+
+    @abstractmethod
+    def _init_pipeline(self, cfg: ConfigType) -> Callable:
+        """Initialize the test pipeline.
+
+        Return a pipeline to handle varies of input data, such as str,
+        np.ndarray. It is an abstract method in BaseInferencer, and should be
+        implemented in subclasses.
+
+        The returned pipeline will be used to process a single data.
+        It will be used in :meth:`preprocess` like this:
+
+        .. code-block:: python
+            def preprocess(self, inputs, batch_size, **kwargs):
+                ...
+                dataset = map(self.pipeline, dataset)
+                ...
+        """
+        raise NotImplementedError('_init_pipeline is not implemented!')
+
+    def _init_visualizer(self, cfg: ConfigType) -> Optional[Visualizer]:
+        """Initialize visualizers.
+
+        Args:
+            cfg (ConfigType): Config contained the visualizer information.
+
+        Returns:
+            Visualizer or None: Visualizer initialized with config.
+        """
+        if 'visualizer' not in cfg:
+            return None
+        timestamp = str(datetime.timestamp(datetime.now()))
+        cfg.visualizer['name'] = f'inferencer-{timestamp}'
+        return VISUALIZERS.build(cfg.visualizer)
+
+    def _get_chunk_data(self, dataset: Iterable, chunk_size: int):
+        """Get batch data from dataset.
+
+        Args:
+            dataset (Iterable): An iterable dataset.
+            chunk_size (int): Equivalent to batch size.
+
+        Yields:
+            list: batch data.
+        """
+        dataset_iter = iter(dataset)
+        while True:
+            try:
+                chunk_data = []
+                for _ in range(chunk_size):
+                    processed_data = next(dataset_iter)
+                    chunk_data.append(processed_data)
+                yield chunk_data
+            except StopIteration:
+                if chunk_data:
+                    yield chunk_data
+                break
+
+    def _dispatch_kwargs(self, **kwargs) -> Tuple[Dict, Dict, Dict, Dict]:
+        """Dispatch kwargs to preprocess(), forward(), visualize() and
+        postprocess() according to the actual demands.
+
+        Returns:
+            Tuple[Dict, Dict, Dict, Dict]: kwargs passed to preprocess,
+            forward, visualize and postprocess respectively.
+        """
+        # Ensure each argument only matches one function
+        for key in kwargs.keys():
+            method_kwargs = [
+                self.preprocess_kwargs,
+                self.forward_kwargs,
+                self.visualize_kwargs,
+                self.postprocess_kwargs,
+            ]
+            matches = tuple(kwarg for kwarg in method_kwargs if key in kwarg)
+            if len(matches) == 0:
+                raise ValueError(
+                    f'unknown argument {key} for `preprocess`, `forward`, '
+                    '`visualize` and `postprocess`')
```

Comment:
method_kwargs are excluded so we only need to check `kwargs.keys()` is the subset of method_kwargs.